### PR TITLE
[Enhancement] Booking of an employee's expense claim

### LINKED ISSUE #7218: Book expense claim against employee account
- [ ] On submission of expense claim, book sanctioned amount in the JL entry against employee account.
- [ ] Set Party Type as Employee
- [ ] Report for Unclaimed Expense Claims
- [ ] Status (paid / unpaid for Expense claim)

### PR DESCRIPTION
**Pull request by rohitwaghchaure**
Fixed #7218

### CODE REVIEW

**Comment by nabinhait** [comment]
@rohitwaghchaure Is it ready to merge?

**Comment by nabinhait** [comment]
test case is failing

**Comment by rohitwaghchaure** [comment]
@nabinhait Travis is fixed

File: erpnext/accounts/doctype/journal_entry/journal_entry.py [hunk 1]
```diff
@@ -653,8 +650,7 @@ def get_payment_entry(ref_doc, args):
 	if args.get("party_account"):
 		# Modified to include the posting date for which the exchange rate is required.
 		# Assumed to be the posting date in the reference document
-		exchange_rate = get_exchange_rate(ref_doc.get("posting_date") or ref_doc.get("transaction_date"), 
-			args.get("party_account"), args.get("party_account_currency"),
+		exchange_rate = get_exchange_rate(ref_doc.posting_date, args.get("party_account"), args.get("party_account_currency"),
```
Comment (nabinhait) [inline erpnext/accounts/doctype/journal_entry/journal_entry.py]: ref_doc.get("posting_date") or ref_doc.get("transaction_date")

File: erpnext/accounts/doctype/journal_entry/journal_entry.py [hunk 2]
```diff
@@ -689,8 +685,7 @@ def get_payment_entry(ref_doc, args):
 		bank_row.update(bank_account)
 		# Modified to include the posting date for which the exchange rate is required.
 		# Assumed to be the posting date of the reference date
-		bank_row.exchange_rate = get_exchange_rate(ref_doc.get("posting_date") 
-			or ref_doc.get("transaction_date"), bank_account["account"], 
+		bank_row.exchange_rate = get_exchange_rate(ref_doc.posting_date, bank_account["account"],
```
Comment (nabinhait) [inline erpnext/accounts/doctype/journal_entry/journal_entry.py]: ref_doc.get("posting_date") or ref_doc.get("transaction_date")

File: erpnext/accounts/doctype/payment_entry/payment_entry.js [hunk 3]
```diff
@@ -26,6 +26,12 @@ frappe.ui.form.on('Payment Entry', {
 			}
 		});
 
+		frm.set_query("party_type", function() {
```
Comment (nabinhait) [inline erpnext/accounts/doctype/payment_entry/payment_entry.js]: set_query not needed

File: erpnext/accounts/doctype/payment_reconciliation/payment_reconciliation.js [hunk 4]
```diff
@@ -6,12 +6,10 @@ frappe.provide("erpnext.accounts");
 erpnext.accounts.PaymentReconciliationController = frappe.ui.form.Controller.extend({
 	onload: function() {
 		var me = this
-		this.frm.set_query('party_type', function() {
-			return {
-				filters: {
-					"name": ["in", ["Customer", "Supplier"]]
-				}
-			};
+		this.frm.set_query("party_type", function() {
```
Comment (nabinhait) [inline erpnext/accounts/doctype/payment_reconciliation/payment_reconciliation.js]: set_query not needed

File: erpnext/accounts/party.py [hunk 5]
```diff
@@ -174,13 +174,13 @@ def get_party_account(party_type, party, company):
 		account = frappe.db.get_value("Party Account",
 			{"parenttype": party_type, "parent": party, "company": company}, "account")
 
-		if not account:
+		if not account and party_type != 'Employee':
```
Comment (nabinhait) [inline erpnext/accounts/party.py]: Use explicit condition `if not account and party_type in ('Customer', 'Supplier')`

File: erpnext/accounts/party.py [hunk 6]
```diff
@@ -174,13 +174,13 @@ def get_party_account(party_type, party, company):
...
 			party_group_doctype = "Customer Group" if party_type=="Customer" else "Supplier Type"
 			group = frappe.db.get_value(party_type, party, scrub(party_group_doctype))
 			account = frappe.db.get_value("Party Account",
 				{"parenttype": party_group_doctype, "parent": group, "company": company}, "account")
 
-		if not account:
+		if not account and party_type != 'Employee':
```
Comment (nabinhait) [inline erpnext/accounts/party.py]: same as above

File: erpnext/accounts/report/general_ledger/general_ledger.js [hunk 7]
```diff
@@ -54,7 +54,7 @@ frappe.query_reports["General Ledger"] = {
 			"fieldname":"party_type",
 			"label": __("Party Type"),
 			"fieldtype": "Select",
```
Comment (nabinhait) [inline erpnext/accounts/report/general_ledger/general_ledger.js]: make it link field, linked to Party Type

File: erpnext/accounts/report/unpaid_expense_claim/unpaid_expense_claim.py [hunk 8]
```diff
@@ -0,0 +1,34 @@
...
+
+	return frappe.db.sql(""" 
+		select
+			ec.employee, ec.employee_name, ec.name, ec.total_sanctioned_amount, ec.total_amount_reimbursed,
+			sum(gle.credit_in_account_currency - gle.debit_in_account_currency) as outstanding_amt
+		from 
+			`tabExpense Claim` ec, `tabGL Entry` gle
+		where
+			gle.against_voucher_type = "Expense Claim" and gle.against_voucher = ec.name
+			and gle.party is not null and ec.docstatus = 1 and {cond} group by ec.name
```
Comment (nabinhait) [inline erpnext/accounts/report/unpaid_expense_claim/unpaid_expense_claim.py]: and ec.is_paid = 0 ???

File: erpnext/hr/doctype/employee/employee.json [hunk 9]
```diff
@@ -2148,6 +2176,34 @@
    "bold": 0, 
    "collapsible": 0, 
    "columns": 0, 
+   "fieldname": "is_frozen",
```
Comment (nabinhait) [inline erpnext/hr/doctype/employee/employee.json]: why needed?

File: erpnext/hr/doctype/employee/employee.json [hunk 10]
```diff
@@ -2148,6 +2176,34 @@
    "bold": 0, 
    "collapsible": 0, 
    "columns": 0, 
+   "fieldname": "is_frozen", 
+   "fieldtype": "Data",
```
Comment (nabinhait) [inline erpnext/hr/doctype/employee/employee.json]: Data?

File: erpnext/hr/doctype/expense_claim/expense_claim.js [hunk 11]
```diff
@@ -176,6 +191,49 @@ frappe.ui.form.on("Expense Claim Detail", {
...
+				}
+			}
+		}
+	},
+
+	set_query_for_payable_account: function(frm) {
+		frm.fields_dict["payable_account"].get_query = function() {
+			return {
+				filters: {
+					"report_type": "Balance Sheet",
```
Comment (nabinhait) [inline erpnext/hr/doctype/expense_claim/expense_claim.js]: Use root_type="Liability" and account_type="Payable"

File: erpnext/hr/doctype/expense_claim/expense_claim.py [hunk 12]
```diff
@@ -21,23 +25,121 @@ def validate(self):
...
 			self.project = frappe.db.get_value("Task", self.task, "project")
 
+	def set_status(self):
+		self.status = {
+			"0": "Draft",
+			"1": "Submitted",
+			"2": "Cancelled"
+		}[cstr(self.docstatus or 0)]
+
+		if self.total_sanctioned_amount == self.total_amount_reimbursed and self.docstatus == 1:
```
Comment (nabinhait) [inline erpnext/hr/doctype/expense_claim/expense_claim.py]: less than equal to

Comment (nabinhait) [inline erpnext/hr/doctype/expense_claim/expense_claim.py]: or is_paid

File: erpnext/hr/doctype/expense_claim/expense_claim.py [hunk 13]
```diff
@@ -21,23 +25,121 @@ def validate(self):
...
+			self.status = "Paid"
+		elif self.docstatus == 1:
+			self.status = "Unpaid"
+
+	def set_payable_account(self):
+		if not self.payable_account and not self.is_paid:
+			self.payable_account = frappe.db.get_value("Company", self.company, "default_expense_payable")
```
Comment (nabinhait) [inline erpnext/hr/doctype/expense_claim/expense_claim.py]: use default_payable_account field

File: erpnext/hr/doctype/expense_claim/expense_claim.py [hunk 14]
```diff
@@ -21,23 +25,121 @@ def validate(self):
...
+				self.get_gl_dict({
+					"account": data.default_account,
+					"debit": data.sanctioned_amount,
+					"debit_in_account_currency": data.sanctioned_amount,
+					"against": self.employee,
+					"cost_center": self.cost_center
+				})
+			)
+
+		if self.is_paid:
```
Comment (nabinhait) [inline erpnext/hr/doctype/expense_claim/expense_claim.py]: If is_paid, make both debit and credit entry for party_account, like POS

File: erpnext/hr/doctype/expense_claim/expense_claim.py [hunk 15]
```diff
@@ -21,23 +25,121 @@ def validate(self):
...
+			# payment entry
+			gl_entry.append(
+				self.get_gl_dict({
+					"account": get_bank_cash_account(self.mode_of_payment, self.company).get("account"),
+					"credit": self.total_sanctioned_amount,
+					"credit_in_account_currency": self.total_sanctioned_amount,
+					"against": ",".join([d.default_account for d in self.expenses]),
+					"against_voucher_type": self.doctype,
```
Comment (nabinhait) [inline erpnext/hr/doctype/expense_claim/expense_claim.py]: against_voucher_type and against_voucher is only used against party, mainly to calculate outstanding, in this case can be used for reimbursement_amount

File: erpnext/hr/doctype/expense_claim/expense_claim.py [hunk 16]
```diff
@@ -64,7 +166,18 @@ def set_expense_account(self):
 		for expense in self.expenses:
 			if not expense.default_account:
 				expense.default_account = get_expense_claim_account(expense.expense_type, self.company)["account"]
-		
+
+def update_reimbursed_amount(doc):
+	amt = frappe.db.sql("""select ifnull(sum(credit_in_account_currency - debit_in_account_currency), 0) as amt 
+		from `tabGL Entry` where against_voucher = %s and against_voucher_type = 'Expense Claim'
+		and party is null """, doc.name ,as_dict=1)[0].amt
```
Comment (nabinhait) [inline erpnext/hr/doctype/expense_claim/expense_claim.py]: party is not null??

calculate based on party and excluding this voucher

File: erpnext/patches/v7_2/update_party_type.py [hunk 17]
```diff
@@ -0,0 +1,24 @@
+# Copyright (c) 2013, Web Notes Technologies Pvt. Ltd. and Contributors
+# License: GNU General Public License v3. See license.txt
+
+from __future__ import unicode_literals
+import frappe
+
+def execute():
```
Comment (nabinhait) [inline erpnext/patches/v7_2/update_party_type.py]: Only make party type for existing accounts.

File: erpnext/setup/doctype/company/company.js [hunk 18]
```diff
@@ -4,6 +4,18 @@
 frappe.provide("erpnext.company");
 
 frappe.ui.form.on("Company", {
+	setup: function(frm) {
+		frm.fields_dict['default_expense_payable'].get_query = function() {
```
Comment (nabinhait) [inline erpnext/setup/doctype/company/company.js]: not needed

File: erpnext/setup/doctype/company/company.json [hunk 19]
```diff
@@ -736,6 +736,35 @@
    "bold": 0, 
    "collapsible": 0, 
    "columns": 0, 
+   "fieldname": "default_expense_payable",
```
Comment (nabinhait) [inline erpnext/setup/doctype/company/company.json]: not needed

File: erpnext/setup/doctype/company/company.py [hunk 20]
```diff
@@ -151,6 +151,9 @@ def set_default_accounts(self):
 			self.db_set("default_income_account", frappe.db.get_value("Account",
 				{"account_name": _("Sales"), "company": self.name}))
 
+		if not self.default_expense_payable:
```
Comment (nabinhait) [inline erpnext/setup/doctype/company/company.py]: not needed

File: erpnext/setup/doctype/party_type/party_type.json [hunk 21]
```diff
@@ -0,0 +1,111 @@
...
+   "email": 1, 
+   "export": 1, 
+   "if_owner": 0, 
+   "import": 0, 
+   "is_custom": 0, 
+   "permlevel": 0, 
+   "print": 1, 
+   "read": 1, 
+   "report": 1, 
+   "role": "Administrator",
```
Comment (nabinhait) [inline erpnext/setup/doctype/party_type/party_type.json]: Administrator?

Give access to Accounts User and Accounts Manager roles

File: erpnext/setup/doctype/party_type/party_type.py [hunk 22]
```diff
@@ -0,0 +1,20 @@
...
+
+from __future__ import unicode_literals
+import frappe
+from frappe.model.document import Document
+
+class PartyType(Document):
+	pass
+
+@frappe.whitelist()
+def get_party_type(doctype, txt, searchfield, start, page_len, filters):
```
Comment (nabinhait) [inline erpnext/setup/doctype/party_type/party_type.py]: not needed

**Review by nabinhait — CHANGES_REQUESTED** [review]

File: erpnext/hr/doctype/expense_claim/expense_claim.py (hunk 12 above)
Comment (rohitwaghchaure) [inline erpnext/hr/doctype/expense_claim/expense_claim.py]: If is_paid is off and user has paid using journal entry

**Pull request merged**